server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -736,6 +742,8 @@ private IndexMetadata(
         this.writeLoadForecast = writeLoadForecast;
         this.shardSizeInBytesForecast = shardSizeInBytesForecast;
         assert numberOfShards * routingFactor == routingNumShards : routingNumShards + " must be a multiple of " + numberOfShards;
+        this.fieldsForModels = fieldsForModels;
+        assert fieldsForModels != null;
```

Comment:
It's going to be critical that `fieldsForModels` is always non-null. I need to review further to ensure there are no code paths that can set it to null.

Reply:
Reviewed more closely and with https://github.com/elastic/elasticsearch/pull/103319/commits/c5748f70ea4ff132567b5cb1b41a0402bd065b45 applied, `fieldsForModels` should always be immutable and non-null.